### ISSUE TITLE
[Performance] How to free GPU memory for transformers ONNX models

### ISSUE THREAD

**Issue by Gaggi72** [microsoft/onnxruntime#19445]
### Describe the issue

I am using the sentence-transformers model with onnx runtime for inferencing embeddings. I have created a FastAPI app on which app startup initialises the Inference session of onnx runtime. Whenever there are new tokens given for embedding creation it occupies GPU memory which is not released after successful execution. I have changed the gpu_mem_limit but still it exceeds it after k iterations. Also, io_bindings are being used for inferencing which get cleared at every API call. How I can free GPU memory

### To reproduce

```
def update_ep_options(provider, model):
        default_ep_options = model.get_provider_options()[provider]
        ep_options = {
          "gpu_mem_limit": "2147483648",
          "arena_extend_strategy": "kSameAsRequested"
        }
        for k, v in ep_options.items():
            default_ep_options[k] = v
        return default_ep_options
```

```
      session_options = onnxruntime.SessionOptions()
      session_options.intra_op_num_threads = 8

      model = onnxruntime.InferenceSession(path_or_bytes=model_path, sess_options=session_options,
                                                providers=['CUDAExecutionProvider'])

      tokenizer = AutoTokenizer.from_pretrained(self.model_name, do_lower_case=True)

      ep_options = update_ep_options(provider='CUDAExecutionProvider', model)
      model.set_providers(['CUDAExecutionProvider'], [ep_options])
      run_options = onnxruntime.RunOptions()
      run_options.add_run_config_entry("memory.enable_memory_arena_shrinkage",
                                            "gpu:0")
```

```
def encode(data: list):
        try:
            inputs = tokenize(data)
            self.bind_inputs_to_device(inputs=inputs)
            start_, end = self.bind_outputs_to_device(input_ids_shape=inputs['input_ids'].shape)
            self.model.run_with_iobinding(self.io_binding, self.run_options)
            embeddings = self.mean_pooling([start_], inputs['attention_mask']).cpu().detach().numpy()
            self.io_binding.clear_binding_inputs()
            self.io_binding.clear_binding_outputs()
            return embeddings

        except Exception as e:
            logger.error(f"Error while encoding-> {traceback.format_exc()} --- {e}")
```

### Platform

Linux

### OS Version

20.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.16.3

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

11.8

### Is this a quantized model?

No

**Comment by pranavsharma** [comment]
This should help. Try setting this run option.
https://github.com/microsoft/onnxruntime/blob/e9ab56fa64c0644a2dc5287d1cd3b945bf7d7981/include/onnxruntime/core/session/onnxruntime_run_options_config_keys.h#L19-L27

**Comment by newgrit1004** [comment]
@pranavsharma, Thank you very much.  It works for me. 

The example python code is here.
[https://github.com/microsoft/onnxruntime/blob/4bfa69def85476b33ccfaf68cf070f3fb65d39f7/onnxruntime/test/python/onnxruntime_test_python.py#L1586](url)

By the way, the inference speed using this option is slower than what I expected. 

Do you have any other options to avoid decreasing inference speed?

**Comment by pranavsharma** [comment]
Shouldn't have impacted the inference speed. Can you tell where the extra time is being spent? Probably in the memory cleanup after the inferencing is complete?

**Comment by hariharans29** [comment]
Some decrease in speed is expected as the memory cleanup involves invoking `cudaFree()` and that cost is cooked into the `Run()` call. To best use this feature, it is important to not allocate weights through the memory pool (arena) and set a high enough "inital" chunk size for the arena such that "most" `Run()` calls do not invoke shrinkage but if there is an outlier `Run()` call that allocates more memory (something that processes large sequnces for example) - that will be freed at the end of the `Run()`. The "initial" chunk is never freed until the end of the session.

Please see detailed coment here - https://github.com/microsoft/onnxruntime/issues/9509#issuecomment-951546580

**Comment by newgrit1004** [comment]
@pranavsharma  yes, it is. @hariharans29 's answer is worth to me. I understood the context. I appreciate that.

**Comment by EvanSong77** [comment]
> @pranavsharma, Thank you very much. It works for me.， 非常感谢。这对我有用。
> 
> The example python code is here.示例 python 代码在这里。 [https://github.com/microsoft/onnxruntime/blob/4bfa69def85476b33ccfaf68cf070f3fb65d39f7/onnxruntime/test/python/onnxruntime_test_python.py#L1586](url)
> 
> By the way, the inference speed using this option is slower than what I expected.顺便说一句，使用此选项的推理速度比我预期的要慢。
> 
> Do you have any other options to avoid decreasing inference speed?您还有其他选择来避免推理速度降低吗？

Thanks! It works for me. I found that performance was not greatly affected.

**Comment by yiluzhuimeng** [comment]
> [@pranavsharma](https://github.com/pranavsharma) yes, it is. [@hariharans29](https://github.com/hariharans29) 's answer is worth to me. I understood the context. I appreciate that.

@hariharans29 @pranavsharma @newgrit1004  i also set run_option.AddConfigEntry("memory.enable_memory_arena_shrinkage", "cpu:0;gpu:0");but while many clients request at the same time , the gpu memory is still bigger, how can you solve it ? can you give me an example of python onnxrumetime-gpu ?
